src/bigtable/mod.rs

```diff
@@ -0,0 +1,594 @@
+    // start/end by value.
+    pub fn with_range(mut self, range: impl RangeBounds<Bytes>) -> Self {
+        let start_key = bound_to_start_key(range.start_bound());
+        let end_key = bound_to_end_key(range.end_bound());
+        let range = v2::RowRange { start_key, end_key };
+        self.row_ranges.push(range);
+        self
+    }
+
+    /// Add a single row to this row set.
+    pub fn with_key(mut self, key: impl Into<Bytes>) -> Self {
+        self.row_keys.push(key.into());
+        self
+    }
+
+    fn restrict_to_after(&mut self, key: Bytes) {
+        self.row_ranges.retain_mut(|r| r.restrict_to_after(&key));
+        self.row_keys.retain(|r| r > &key);
+    }
+}
+
+/// The default [`RetryPredicate`] used for errors from Bigtable operations
+#[derive(Debug, Default, Clone)]
+pub struct BigtableRetryCheck {
+    _priv: (),
+}
+
+impl BigtableRetryCheck {
+    /// Create a new instance with default settings
+    pub fn new() -> Self {
+        Self { _priv: () }
+    }
+}
+
+impl RetryPredicate<tonic::Status> for BigtableRetryCheck {
+    fn is_retriable(&self, error: &tonic::Status) -> bool {
+        use tonic::Code;
+
+        // this error code check is based on the ones used in the Go bigtable client lib:
+        // https://github.com/googleapis/google-cloud-go/blob/66e8e2717b2593f4e5640ecb97344bb1d5e5fc0b/bigtable/bigtable.go#L107
+
+        match error.code() {
+            Code::DeadlineExceeded | Code::Unavailable | Code::Aborted => true,
+            _ => false,
+        }
+    }
+}
+
+pub use api::bigtable::v2::{Cell, Column, Family, Row};
+
+/// A reference for a single cell in Bigtable
+pub struct CellRef<'a> {
+    /// The column family name
+    pub family_name: &'a str,
+    /// The column qualifier
+    pub column_qualifier: &'a prost::bytes::Bytes,
+    /// The cell's timestamp, in microseconds since the epoch
+    pub timestamp_micros: i64,
+    /// The cell's contents
+    pub value: &'a prost::bytes::Bytes,
+}
+
+impl Row {
+    /// Iterate over this row's cells, returning only the most recent cell in each column
+    pub fn most_recent_cells(&self) -> impl Iterator<Item = CellRef<'_>> + '_ {
+        self.families.iter().flat_map(|fam| {
+            fam.columns.iter().filter_map(move |col| {
+                // The cells vec is *decreasing* in `timestamp_micros`
+                col.cells.first().map(move |cell| CellRef {
+                    family_name: &fam.name,
+                    column_qualifier: &col.qualifier,
+                    timestamp_micros: cell.timestamp_micros,
+                    value: &cell.value,
+                })
+            })
+        })
+    }
+}
+
+#[derive(Default)]
+struct ReadInProgress {
+    row: Row,
+    value: BytesMut,
+}
+
+impl ReadInProgress {
+    fn flush_bytes_in_progress(&mut self) {
+        if !self.value.is_empty() {
+            let value = self.value.split().freeze();
+            if let Some(cell) = self.last_cell_mut() {
```

Comment:
Hey @rnarubin - I'm picking this up from @jneem to try & get it merged. 

As far as I can tell, what you're saying here is correct - `self.families` is only empty when the row is empty (in which case `self.value` should also be empty) ([source](https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2#row)), and if we have a `family` then `family.columns` is never empty ([source](https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2#google.bigtable.v2.Family)).

As a result, I'm now wondering what strength of assertion we need here. We could have each of these methods return a `Result` & update `BigtableClient`'s `read_rows` method to return `impl Stream<Item = Result<Row, SomeNewErrorType>>` - which is what I think you're proposing - but I'm not really sure how a client is meant to respond to an error that essentially says "we've ended up in an inconsistent state". Essentially, I'm wondering whether it might be valid to handle this with `panic!` or `unreachable!` (ideally combined with the conformance tests you mentioned above, which I'd like to work on), rather than bubbling it up to the consumer. What do you think?

Reply:
Yes i think an error is the right call instead of a panic, though acknowledging that this error isn't really actionable. I'd consider panics as conventionally "our software reached an impossible state/did something wrong", but here the cause would be a misbehaving server -- or more generally, bad network inputs. An error feels more appropriate when the cause is out of our control.

If the user really wants, they could conceivably retry. A panic is more terminal